### DISCUSSION TITLE
Option to share collections

### DISCUSSION THREAD

**Discussion by Phippe** [ppy/osu#13025] [Ideas]
I'd like to see a share option for collections, so you could send something like an osk-file to a friend and they'll have the same collections (as long as the corresponding maps are available in their library).

**Comment by alikindsys** [comment]
Any idea on what could be the best way to implement this? What would be the new file extension and would it be a custom file format or a JSON object that contains the data of a given `BeatmapCollection`? 

Cause what i've seen so far all thats needed is to store the name and an array of beatmap ids.
```json
{
    "Name":"My fancy collection",
    "Beatmaps":[1234,5678,9101112]
}
```
That later can just be fetched from either the current user library or from osu!direct, if the maps aren't available.

**Comment by Tresquel** [comment]
What about just uploading the collection to the server and just sending a friend the collection? Like a button to "Share to friends" or something like that

**Comment by peppy** [comment]
Uploading collections is a completely different feature, which I believe is already being tracked on `osu-web`.

**Reply by Akarinnnnn** [reply]
Considering such use case, a I'm sharing a collection privately to my friend. In this time, we can just provide my friend a pretty serialized JSON array containing collection data. Furthermore, this file can have an extension named `.osc.json` stands for "osu! Collection in JSON format".

Using `System.Text.Json.JsonSerializer.SerializeAsync/DeserializeAsync<ExportedCollectionItem[]>(Stream, ...)` is enough for this solution. If you want this feature, I would like to help implement it.

But for the UI/UX and behavior after imported this file, I admit we need more consideration. For example how to handle those map and set which is not present yet; For missing online information whether should we fetching them back and bring it to UI(maybe popup); Should we add a identifier(GUID or something similiar) to avoid importing same collection multiple times? Or let user choose creating new otherwise merge to existing one? 

I'm not good at designing UI/UX, so for this topic I can only provide a simple prototype, here is it:
<img width="998" height="640" alt="collection importing popup" src="https://github.com/user-attachments/assets/23c43177-067f-4a41-998f-a22944e53129" />

**Comment by 20Attila03** [comment]
I would also add that when the player download someone's collection. And doesn't have the maps. The player (if supporter) would have the option to download all the missing maps from the set and apply them to it.

**Comment by developomp** [comment]
This would also be helpful when reinstalling osu

**Reply by peppy** [reply]
You could just keep a backup of your files for this purpose.

**Reply by 20Attila03** [reply]
I feel like, if you wanna play at your friend's place, it could help downloading your favorite maps, otherwise just do a backup.

**Reply by MishterKirby** [reply]
>
> 
> You could just keep a backup of your files for this purpose.

Would be helpful to have something like this suggestion for some scenarios, though. (My hard drive died the other day and I had to manually hunt down all my beatmaps again, because I didn't have anywhere else to put them)

**Reply by developomp** [reply]
My biggest reason for supporting this feature is because it can be used to sync beatmaps between lazer and stable.

**Reply by peppy** [reply]
> My biggest reason for supporting this feature is because it can be used to sync beatmaps between lazer and stable.

no it couldn't. collections require beatmaps to already be present. you are asking for much more.

**Comment by EnergoStalin** [comment]
I think the ultimate solution would be to have [#13194](https://github.com/ppy/osu/discussions/13194#discussion-3384591), [#2351](https://github.com/ppy/osu/issues/2351#issuecomment-378777154) combined in something like this.

- Export collection(s) from whatever menu into let's say .osc file
    Minimal data for search should be exported (ie. bmid, title, mapper, bpm, star rate, hash?)
- ...
- Import collection from the dedicated osu site page or from file
    Import from file may be extremely useful to maintain local convertibility with other collection formats (ie. [osdb](https://github.com/Piotrekol/CollectionManager/blob/8217d3c9baefa0335e8d5e2b3bb43b870c4a291b/CollectionManagerDll/Modules/FileIO/FileCollections/OsdbCollectionHandler.cs#L18)) at least from my view
- Collection creates from .osc linking against existing beatmaps
- Beatmaps not presented locally grayed out
    Checking against site not performed for them until download of course
- To download map just double click it
    If automatic download fails beatmap browser opens with \<artist> - <title> \<mapper> prompt

**Reply by Tenexxt** [reply]
This would be extremely helpful for sharing mappools quickly, since currently the best solution is putting all .osz files in a .zip, or just downloading everything by yourself.

Hopefully functionality like this will be added someday

**Reply by tadatomix** [reply]
Sounds like an edge case, however there should be something to resolve the issue with local-only beatmaps in collection. Or beatmaps that were in graveyard but later were deleted. I think for both cases, both parties (exporter and importer) should be informed about this (exporter needs to know that some beatmaps (list which one) would not be downloaded by importers and importer need to know what beatmaps would not be imported and maybe why that might happen)

**Reply by Tenexxt** [reply]
> Sounds like an edge case, however there should be something to resolve the issue with local-only beatmaps in collection.

For maps that are detected to be local-only, the .osc file could store the .osz data with a warning that the file size of the .osc would increase because of that. I think it would be fine because this would practically be used for Beatmap packs and mappools -which have their maps uploaded to osu! like 99% of the time. Others would probably just use this to share their favorite beatmaps in a single file, which are also uploaded maps 90% of the time

> Or beatmaps that were in graveyard but later were deleted.

Since .osc files (or any files in general) aren't sentient and can't update their data out of the blue to include the deleted maps on the site i think it would be the best for the game to give a `[X] beatmap canot be found in the osu! beatmap listing, please ask the person that shared this pack to send the map seperately` type warning (this could also appear when already local-only maps were listed in the collection but could not be found and the .osc wouldn't have the map built into it)

I think this shoudn't be done by uploading to the osu site as a seperate thing similary to .osz files,it would work better as just a another file you can download from various places like .osk. But if there is any place on the site that could benefit from the .osc its the beatmap packs 100%, having to always add them to the collection manually gets annoying quickly

**Reply by Akarinnnnn** [reply]
Showing this to importers? 
<img width="998" height="640" alt="image" src="https://github.com/user-attachments/assets/9147e9b6-88ee-4aa8-86b5-aed9ec373a63" />

**Reply by EnergoStalin** [reply]
I remember peppy saying that serving beatmapsets is costly so i doubt that automatic mindless download option would be an option in official clients. Think why there is no such thing in already existing for similar purpose playlists?

That's why i stated that best compromise would be to allow "ghost beatmaps" in listing or integrate 3rd party mirrors(which osu already has procedure of integrating as official ones so i doubt that something like that would be ever avaliable). Even if so 3rd party mirrors have it's own problems since maps still should be validated against bancho to guarantee it's latest version.

So first i would suggest looking into "ghost beatmaps" for online beatmap case which require considerable amount of work for integrating into existing listing and probably cause a lot of confusion. Though decoupling beatmap metadata from on disk data would be extremely cool.

Or just teach lazer to import zips with nested oszs as collections for now. Which is most easy way to cover mappool sharing usecase.

**Reply by Akarinnnnn** [reply]
Another edge case:
>  Should we add a identifier(GUID or something similiar) to avoid importing same collection multiple times? Or let user choose creating new otherwise merge to existing one?

**Reply by Akarinnnnn** [reply]
> Or just teach lazer to import zips with nested oszs as collections for now. Which is most easy way to cover mappool sharing usecase.

I imagine `.osc` has such structure. For missing maps can we include it like this? Or even include all maps?
```
COLLECTION.OSC
    2016343 QUEENDOM - Ciki Ciki Bam Bam.osz
    collection.json
```

**Reply by EnergoStalin** [reply]
Even if "download all" probably wouldnt be approved unless you convince developers otherwise. imo you still can try going for osc. Even might reuse component for playlist listing at least allowing to import embedded maps along with shown online collection listing for standard click & load.

Also cool trick probably would be to create 0 size archive entries instead of collection.json to avoid unneccesary nesting if you dont plan to store whole metadata of course.

In case i got you wrong and collection.json is not for online maps but cache of embedded ones likely it wont be needed at all. Since i'm pretty sure you can observe import progress getting metadata from it and populating collection that way.

**Reply by Akarinnnnn** [reply]
> Even if "download all" probably wouldnt be approved unless you convince developers otherwise. imo you still can try going for osc. Even might reuse component for playlist listing at least allowing to import embedded maps along with shown online collection listing for standard click & load.

Your concern is right。In my inmagine it shares map item from multiplayer. I know less about online.collection so I skip it. Let's talk about .osc, embedding maps inside is necessary in my option, since there are ghost maps and poor connected players.

> Also cool trick probably would be to create 0 size archive entries instead of collection.json to avoid unneccesary nesting if you dont plan to store whole metadata of course.

If this trick is used, it will obviously make pressure to metadata server. I concern devs don't like this. This is why I store cached MD(metadata) in collecrions.json

> In case i got you wrong and collection.json is not for online maps but cache of embedded ones likely it wont be needed at all. Since i'm pretty sure you can observe import progress getting metadata from it and populating collection that way.

Another reason that cache MD in json is, when we meet ghost maps, we can still have a beatmap preview without having to import it. This is acheived by chched MD. Also, for players having poor connection to server(mostly in China, excluding SARs and Taiwan province), if they can't fetch MD from server, cached MD in json will save their lives.

**Reply by EnergoStalin** [reply]
Let's clear what i ment by using this self given titles:

*preview - background
*online map - metadata only (title, hashes, author, ...)
*offline map - full osz
*ghost map - listed locally online map (metadata for preview/download but no osz)

Can't not make honorable mention of [CollectionManager](https://github.com/Piotrekol/CollectionManager) which already excelled at this as 3rd party tool a long long time ago. Would be very cool to have at least some features upstream. The whole online/offline concept was practically inherited from it.

> online.collection

By online collection i mean existing playlist view which i suspect is the same as multiplayer one.

> embedding maps inside is necessary in my option

Agreed for offline maps it is necessary but for online(which listed on bancho) should be an option? If you plan to support slim collections which will have online maps.

> If this trick is used, it will obviously make pressure to metadata server.

Not neccessarily since we cant guarantee bg being avaliable for online maps unless it's base64 encoded during generation but more likely with 0 sized entries. I thought that all hard required information for displaying `<artist> - <title> [<diff>] <author> <diff hash>` could be encoded in filename if really needed.

Also don't think devs would be against it considering currently whole playlist vew solely rely on online metadata for new maps and maybe some local caching as well as beatmap browser do. Metadata is cheap if you dont try to load 2k beatmaps at once lazer basically lazy load everything already.

<details>
<summary>Layout proposal</summary>

```
COLLECTION.<OSC/zip/7z(doesent matter as long as it opens as archive)>
	<id> <artist> - <title>.osz # Beatmapset
	<id> <artist> - <title> [<diff hash>].json
	<id> <artist> - <title>.jpg (beatmapset preview?)
	<id> <artist> - <title> [<diff hash>].jpg (beatmap preview?) # Some maps in set can have different bg
```

That way we can make effective assumptions and implicitly support many combinations without storing extra data while also providing transparency to archivation tools.

For example:

1. If no .json associated with .osz add whole beatmapset to collection.
2. If no .jpg associated with .osz ignore preview

Now we support all offline archives without metadata at once means all existing beatmap packs/tournament pools(except exact diffs being tracked could be improved as adoption grows).

Inside of `*.json`
```json
{
		"title": "<unicode title>(<artist> - <title> <diffname>)",
		"artist": "<unicode artist>",
		"author": "...",
		"preview": "<base64 encoded preview?>"
}
```

Though layout looks minimal and great on paper manually restoring framework object model from it could be not that straightforward and error prone. So caching let's say whole APIBeatmap may look more appealing if it's avaliable at the place.

My main motivation for using separate files for beatmap/metadata/preview is interopability with standard tools like `7z l COLLECTION.OSC` without need to know internal json layout.

</details>

> when we meet ghost maps, we can still have a beatmap preview without having to import it

For offline maps you still technically can pre parse bg from osz inside osc as for online there only embedding can help which i strongly recommend having option to strip away as this likely wont be needed most of the time for the "quick share".

> for players having poor connection to server(mostly in China, excluding SARs and Taiwan province)

F russia next. Sometimes minor inconsitencies with bm.ppy.sh can be noticed even in least least restricted region i live in. Still considering writing reverse proxy which will locally balance traffic between mirrors. So bancho wont be a single point of falure anymore. Wont help with metadata though...

**Reply by Akarinnnnn** [reply]
> Let's clear what i ment by using this self given titles:
> 
> - preview - background 
> - online map - metadata only (title, hashes, author, ...) 
> - offline map - full osz 
> - ghost map - listed locally online map (metadata for preview/download but no osz)

1. storing metadata only is absolutely works for ranked maps, but if we store other online maps like this, ghost map may present.
2. i agree storing full osz for locals inside

> > online.collection
> 
> By online collection i mean existing playlist view which i suspect is the same as multiplayer one.
Sorry it's a typo, but thsnks for sharing yoir design.
> > embedding maps inside is necessary in my option
> 
> Agreed for offline maps it is necessary but for online(which listed on bancho) should be an option? If you plan to support slim collections which will have online maps.
> 
> > If this trick is used, it will obviously make pressure to metadata server.
> 
> Not neccessarily since we cant guarantee bg being avaliable for online maps unless it's base64 encoded during generation but more likely with 0 sized entries. I thought that all hard required information for displaying `<artist> - <title> [<diff>] <author> <diff hash>` could be encoded in filename if really needed.
> 
> Also don't think devs would be against it considering currently whole playlist vew solely rely on online metadata for new maps and maybe some local caching as well as beatmap browser do. Metadata is cheap if you dont try to load 2k beatmaps at once lazer basically lazy load everything already.
> 
> Layout proposal
> > when we meet ghost maps, we can still have a beatmap preview without having to import it
> 
> For offline maps you still technically can pre parse bg from osz inside osc as for online there only embedding can help which i strongly recommend having option to strip away as this likely wont be needed most of the time for the "quick share".

For pre parsing, it need to run expensive importer, but result won't get cached instead formally imported at current status iirc. I have performance and behaviorial inconsistent concernings about this route. Alternatively, i have an idea, embed preview bg in /bg/{map.Md5} without extension, audio is liked. Of course, all I wrote about is mostly for potentional ghost maps or local ones. Through apply this logic on ranked maps can also being benifited by instantly got preview after `.osc` is ready for UI.

> > for players having poor connection to server(mostly in China, excluding SARs and Taiwan province)
> 
> F russia next. Sometimes minor inconsitencies with bm.ppy.sh can be noticed even in least least restricted region i live in. Still considering writing reverse proxy which will locally balance traffic between mirrors. So bancho wont be a single point of falure anymore. Wont help with metadata though...

:( good luck to you. So do you think embed metadata is a good idea? Through it can't encoded in filename directly sometimes (because titie may contain `\` or `/`, or filename too long for zip), we can still store it in manifest.

**Reply by Akarinnnnn** [reply]
> Let's clear what i ment by using this self given titles:
> 
> - preview - background 
> - online map - metadata only (title, hashes, author, ...) 
> - offline map - full osz 
> - ghost map - listed locally online map (metadata for preview/download but no osz)

1. storing metadata only is absolutely works for ranked maps, but if we store other online maps like this, ghost map may present.
2. i agree storing full osz for locals inside
3. I defined "pre-import maps" for maps that is info loaded for UI and not being download or present on disk. Is this your ghost map mean?
4. Define another term dead maps, means metadata only maps which can't retrieve file by various reasons, like new version is oublished.

> > online.collection
> 
> By online collection i mean existing playlist view which i suspect is the same as multiplayer one.

Sorry it's a typo, but thsnks for sharing your design.

> > embedding maps inside is necessary in my option
> 
> Agreed for offline maps it is necessary but for online(which listed on bancho) should be an option? If you plan to support slim collections which will have online maps.

I think for silm mode, what we can trim is only ranked(approved maybe?) maps, because they are immutable. Other type of maps may change then lead to dead maps in a collection.

> > If this trick is used, it will obviously make pressure to metadata server.
> 
> Not neccessarily since we cant guarantee bg being avaliable for online maps unless it's base64 encoded during generation but more likely with 0 sized entries. I thought that all hard required information for displaying `<artist> - <title> [<diff>] <author> <diff hash>` could be encoded in filename if really needed.

I don't think we can do this, path seperators `/\` and envvar seperstors `:;` in these field may put this file in incorrect place or even cause import failure(DoS), this is why i want to put them in manifest file.

> Also don't think devs would be against it considering currently whole playlist vew solely rely on online metadata for new maps and maybe some local caching as well as beatmap browser do. Metadata is cheap if you dont try to load 2k beatmaps at once lazer basically lazy load everything already.
> 
> Layout proposal
> > when we meet ghost maps, we can still have a beatmap preview without having to import it
> 
> For offline maps you still technically can pre parse bg from osz inside osc as for online there only embedding can help which i strongly recommend having option to strip away as this likely wont be needed most of the time for the "quick share".

For pre parsing, it need to run expensive importer, but result won't get cached instead formally imported at current status iirc. I have performance and behaviorial inconsistent concernings about this route. Alternatively, i have an idea, embed preview bg in /bg/{map.Md5} without extension, audio is liked. Of course, all I wrote about is mostly for potentional dead maps or local ones. Through apply this logic on ranked maps can also being benifited by instantly got preview after `.osc` is ready for UI.

> > for players having poor connection to server(mostly in China, excluding SARs and Taiwan province)
> 
> F russia next. Sometimes minor inconsitencies with bm.ppy.sh can be noticed even in least least restricted region i live in. Still considering writing reverse proxy which will locally balance traffic between mirrors. So bancho wont be a single point of falure anymore. Wont help with metadata though...

:( good luck to you. So do you think embed metadata is a good idea? Through it can't encoded in filename directly sometimes (because titie may contain `\` or `/`, or filename too long for zip), we can still store it in manifest.

**Reply by Akarinnnnn** [reply]
if we want filenamed metadata, use json format like this is easier for machine reading then do more processing for it: `[1, <artist>, <original artist>, <title>, <original title>, <diff>, <author>, <diff hash>, <SR>]`. In this format, users prefer original language won't get break.

1 stands for version. For better compatbility we can encode it into a json object instead of array.

**Reply by EnergoStalin** [reply]
I would set such goals for now.

1. Preserve exact .osu version at the time of creating collection.
2. Store enough metadata for restoring the rest of osz in case of loss cause other files is just assets for the map.
3. Should support beatmap packs(plain zip with oszs inside).
4. Should not compromise on transparency without bringing useful features.
5. Should support slim collections which work for online as well as offline maps.
6. Should provide sensible audio/thumbnail preview for offline maps (?).

> storing metadata only is absolutely works for ranked maps, but if we store other online maps like this, ghost map may present.

Not quite i wouldnt rely on map being ranked since i remember some of them may get unranked during reevaluation. In practice extremely rare but still possible.

> For pre parsing, it need to run expensive importer, but result won't get cached instead formally imported at current status iirc. I have performance and behaviorial inconsistent concernings about this route.

If you do it through lazer importer sure it expensive. Though with custom one locally unzipping a couple of megabytes even from nested archives shouldnt be a problem for a pc lazer can stabily run on. imo with lazyloading latency is negligeble for this use case. You wouldnt sit scrolling the view all day do you?

> embed preview bg in /bg/{map.Md5} without extension, audio is liked

Audio is tough will you plan on ripping preview point of the song compressing it? It should also work for all local only maps for which preview point audio segment unavaliable online.

Nested dir for bg is a good idea. Though still 2 variants PanelBackground and BeatmapCardThumbnail needed for PlaylistItem.
https://github.com/ppy/osu/blob/61ecd18d2491697c77ef73f47d02472c6b7ed5a2/osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs#L96-L98

Also looks like stock DrawableRoomPlaylistItem cant be injected with beatmap data. It always rely on OnlineID being present.
https://github.com/ppy/osu/blob/61ecd18d2491697c77ef73f47d02472c6b7ed5a2/osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs#L188

> So do you think embed metadata is a good idea?

Now i think it's the only choice. Cant rely on .osu files for online linking throughout versions.

> Through it can't encoded in filename directly sometimes

Idea was from how legacy client Songs folder structured. So not without escaping. Though it's really old and inconvinient approach.

> I defined "pre-import maps" for maps that is info loaded for UI and not being download or present on disk. Is this your ghost map mean?

Yes

> Define another term dead maps, means metadata only maps which can't retrieve file by various reasons, like new version is oublished.

Could be partially mitigated by storing .osu file.

> `[1, <artist>, <original artist>, <title>, <original title>, <diff>, <author>, <diff hash>, <SR>]`

Dont know about that. Fields probably still should be escaped for `\/` `:;` since json wont handle it.

Also as an alternative approach maybe throw lazer realm to .osc for metadata with related files like current storage driver do? That way it would be deduplicaed though make merging/managing impossible without specialized tool. Importing would be harder too.

**Reply by Akarinnnnn** [reply]
6&5 could share code if we use unified "on demand" storing osz policy, with metadata of collection always present, if you like.

But I'm on work and too busy to share details. Don't worry I will be available at 19:30 UTC+8.
